samples/cxx/jacobian/jacobian_speed.cpp

```diff
@@ -0,0 +1,249 @@
+#include <iomanip>
+#include <numeric>
+#include "cantera/base/Solution.h"
+#include "cantera/thermo/IdealGasPhase.h"
+#include "cantera/kinetics.h"
+#include "cantera/numerics/eigen_sparse.h"
+
+using namespace Cantera;
+
+void statistics(vector_fp times, size_t loops, size_t runs)
+{
+    double average = accumulate(times.begin(), times.end(), 0.0) / times.size();
+    for (auto& v : times) {
+        v = (v - average) * (v - average);
+    }
+    double std = accumulate(times.begin(), times.end(), 0.0) / times.size();
+    std = pow(std, 0.5);
+
+    // output statistics
+    std::cout << std::setprecision(5) << average / 1000. << " μs ± "
+        << std::setprecision(3) << std / 1000. << " μs "
+        << "per loop (" << runs << " runs, " << loops << " loops each)\n";
+}
+
+//! timer for standard getters
+void timeit_base(void (Kinetics::*function)(double*),
+                 Kinetics* kin,
+                 ThermoPhase& gas,
+                 size_t siz,
+                 size_t loops=10000,
+                 size_t runs=7)
+{
+    vector_fp out(siz);
+
+    double T = gas.temperature();
+    double pressure = gas.pressure();
+    double deltaT = 1e-5;
+
+    vector_fp times;
+    for (size_t run = 0; run < runs; ++run) {
+        auto t1 = std::chrono::high_resolution_clock::now();
+        for (size_t i = 0.; i < loops; ++i) {
+            gas.setState_TP(T + i * deltaT, pressure);
+            (kin->*function)(out.data());
+        }
+        auto t2 = std::chrono::high_resolution_clock::now();
+        times.push_back(
+            std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1).count() /
+            loops);
+    }
+    gas.setState_TP(T, pressure);
+    statistics(times, loops, runs);
+}
+
+//! timer for Eigen::VectorXd getters
+void timeit_vector(Eigen::VectorXd (Kinetics::*function)(),
+                   Kinetics* kin,
+                   ThermoPhase& gas,
+                   size_t loops=10000,
+                   size_t runs=7)
+{
+    Eigen::VectorXd out;
+
+    double T = gas.temperature();
+    double pressure = gas.pressure();
+    double deltaT = 1e-5;
+
+    vector_fp times;
+    for (size_t run = 0; run < runs; ++run) {
+        auto t1 = std::chrono::high_resolution_clock::now();
+        for (size_t i = 0.; i < loops; ++i) {
+            gas.setState_TP(T + i * deltaT, pressure);
+            out = (kin->*function)();
+        }
+        auto t2 = std::chrono::high_resolution_clock::now();
+        times.push_back(
+            std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1).count() /
+            loops);
+    }
+    gas.setState_TP(T, pressure);
+    statistics(times, loops, runs);
+}
+
+//! timer for Eigen::SparseMatrix<double> getters
+void timeit_matrix(Eigen::SparseMatrix<double> (Kinetics::*function)(),
+                   Kinetics* kin,
+                   ThermoPhase& gas,
+                   size_t loops=1000,
+                   size_t runs=7)
+{
+    Eigen::VectorXd out;
```

Comment:
It looks like this function assigning a `SparseMatrix<double>` to a `VectorXd`? What does that do?

Reply:
Good catch! It's just used to receive the output value. Fixing it doesn't appear to change much for benchmarks, thankfully.